### PR TITLE
Attempt to add postgres metrics via postgres_exporter

### PR DESCRIPTION
**Pull request by idoqo**
The approach I'm currently taking here is to:
- have a generic "DockerService" struct that can be used as a template for docker services we need to run
- Setup a "PostgresExporterService" with all the info needed to run postgres_exporter as a service
- Start the PostgresExporterService on startup*

There are some cleanups needed here e.g
- We should only start the exporter service if there's an existing PostgreSQL cluster. Otherwise, we should start the service when the user creates a new cluster. 
_Alternatively - we can watch for a new cluster and start the exporter service once we detect a PG cluster._

- We need to ensure all services started are in the same network so that they can communicate.
- Maybe we can provide a way for users to disable monitoring?
- The PGExporterService only shows metrics from the PostgreSQL - we will have to provide a Prometheus config where this service (and the API) are [scrape targets](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config) so that the user can view all of them in one place

### CODE REVIEW

**Comment by idoqo** [comment]
Hi @viggy28, I did some learning like we discussed and here's some of what I learnt on docker-compose vs docker

### Stopping a running service
I think stopping running services is the same whether we use `docker-compose` or just `docker`. 

**Docker**: `docker stop $container`.
**docker-compose**: `docker-compose stop $service`.

In this case, I feel like they both are equally hard (or easy) I guess :).

Only problem though, is that removing a service from docker-compose.yml and re-running docker-compose up doesn't remove the container (if it's running).

### Starting a new service on startup
With docker-compose, we'd declare all the needed services in the docker-compose file and running `docker-compose up` will start them.

With docker, we'd have to create each container and start them up ourselves (we'd need to keep track of the container name though).

I feel main advantage of docker-compose here is the ease (it automatically knows the services we created).

The disadvantage here is that if one service fails, all other services will be started (but the command returns an error ) so we might not be able to tell. It is also harder to tell which of the services failed (aside logs) and why it failed.

### Starting a new service on demand
With docker-compose, we'd have to update `docker-compose.yml` (or any other docker-compose file) to start a new service.

For docker, we can create the service on the fly and start it.

I think docker-compose has some disadvantage in this case, since we will have to properly update the docker-compose file.

**Comment by viggy28** [comment]
Thanks, @idoqo for the detailed explanation.

> Only problem though, is that removing a service from docker-compose.yml and re-running docker-compose up doesn't remove the container (if it's running).

That means, do we need to manually remove the container?

Also, what about connectivity (networking) between containers? IIRC, between all the containers created by a docker-compose service, it works out of the box.

**Comment by idoqo** [comment]
Yeah, we will have to remove the container manually as in the "stopping a running service" part. 

I think networking isn't a big problem like I initially thought - `docker-compose` automatically creates its network. If we use regular docker, we can also simply create a network on startup, and have all services join that network. Either way, docker will manage the network for us

**Comment by viggy28** [comment]
Well, then I am fine with either of the approaches. Go for it whichever you feel confident on :)

**Review by viggy28 — CHANGES_REQUESTED** [review]

File: internal/dockerservice.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+		Name:          "postgres_exporter",
+		NetworkName:   networkName,
+		RestartPolicy: "unless-stopped",
+		Ports: map[int]int {
+			9187: 9187,
+		},
+		Environment: map[string]string {
+			"DATA_SOURCE_NAME": fmt.Sprintf("postgresql://%s:%s@%s:5432/postgres?sslmode=disable",
+				"postgres",
+					"replaceme",
```
[the remark this answers is not in the document]
Comment (idoqo) [reply internal/dockerservice.go]: Yeah, for now the username and passwords are stored in sqlite, so we just pull those and construct the dsn here

File: internal/dockerservice.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+		log.Print(err)
+		return err
+	}
+	return nil
+}
+
+func NewPgExporterService(networkName string) DockerService {
+	exporterSvc := DockerService{
+		Name:          "postgres_exporter",
+		NetworkName:   networkName,
```
[the remark this answers is not in the document]
Comment (viggy28) [reply internal/dockerservice.go:75]: I think so.

File: internal/dockerservice.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	args = append(args, ds.Image)
+
+	return args
+}
+
+func (ds DockerService) Start() (err error) {
+	logger := zerolog.New(os.Stderr).With().Str("name", ds.Name).Timestamp().Logger()
+	_, err = ds.DockerClient.NetworkCreate(context.TODO(), ds.NetworkName, types.NetworkCreate{CheckDuplicate: true})
+	if err != nil {
+		networkExistsError := fmt.Sprintf("network with name %s already exists", ds.NetworkName)
```
[the remark this answers is not in the document]
Comment (viggy28) [reply internal/dockerservice.go:55]: smart.

File: api/create.go [hunk 4]
```diff
@@ -145,9 +150,24 @@ func startService(s service, path string) error {
...
+		return "", err
 	}
-	return nil
+	serviceContainerID, err = lastContainerID()
+
+	if s.Db.Monitoring == "enable" {
+		cli, err := client.NewClientWithOpts(client.FromEnv)
+		if err != nil {
+			fmt.Println(err)
+			return serviceContainerID, err
```
[the remark this answers is not in the document]
Comment (viggy28) [reply api/create.go:166]: makes sense. that's right.

File: README.md [hunk 5]
```diff
@@ -112,40 +112,46 @@ postgres=# \dt
...
-
-- Method:
+#### Method:
 
 `POST`
 
-- Data Params
-
-```
+#### Data Params
```
[the remark this answers is not in the document]
Comment (viggy28) [reply README.md:123]: that's fair.

**Pull request merged**